### PR TITLE
incremental translation implemented

### PR DESCRIPTION
**Pull request by confile**
translation task now:

- detects source file changed in source files and generated source files, it only translates source files which have been changed since the last compile 
- removes .java source files and remove translated counterpart .m and .h files 

This pretty cool and makes development a lot faster.

### CODE REVIEW

**Comment by confile** [comment]
@brunobowden May be you have an idea how to add the .class files correctly to the classpath here https://github.com/confile/j2objc-gradle/commit/af138abd9a986c2c27b895d6193dfed0cbf79653#diff-cd56e928e1096973b671c75e5e0c6073R525

When I uncomment the code I get these output messages: 
```
warning: error in opening zip file
```

File: j2objc.gradle [hunk 1]
```diff
@@ -417,8 +418,8 @@ class J2objcCycleFinderTask extends DefaultTask {
         }
     }
 }
-
-
+     
+
```
Comment (brunobowden) [inline j2objc.gradle]: Avoid adding spaces. Same for previous line as well.

File: j2objc.gradle [hunk 2]
```diff
@@ -426,7 +427,48 @@ class J2objcTranslateTask extends DefaultTask {
     @OutputDirectory File destDir
 
     @TaskAction
-    def translate() {
+    def translate(IncrementalTaskInputs inputs) {
+        println "Source files: "+srcFiles.getFiles().size()
+        FileCollection srcFilesChanged = project.files()
+        inputs.outOfDate { change ->
+        //  println "New or Updated file: "+change.file
+          srcFilesChanged += project.files(change.file)
```
Comment (brunobowden) [inline j2objc.gradle]: Should be 4 space tabs. Same issues repeats several times in the task.

File: j2objc.gradle [hunk 3]
```diff
@@ -434,12 +476,14 @@ class J2objcTranslateTask extends DefaultTask {
...
         // Clean build directory to delete files that are no longer generated
         // Print command before operation to be helpful log if operation fails
-        println "Deleting j2objc build dir: " + destDir.path
-        project.delete destDir.path
-
+        if (freshBuild) {
+          println "Deleting j2objc build dir: " + destDir.path
+          project.delete destDir.path
+        }
+
```
Comment (brunobowden) [inline j2objc.gradle]: Remove "freshBuild" logic, this shouldn't be needed and it's better to have a single code path. If you have a rarely used path like this, it invites bugs that will only be rarely seen.

File: j2objc.gradle [hunk 4]
```diff
@@ -449,7 +493,7 @@ class J2objcTranslateTask extends DefaultTask {
         }
 
         // Generated Files
-        srcFiles = J2objcUtils.addJavaFiles(project, srcFiles, project.j2objcConfig.generatedSourceDir)
+       // srcFiles = J2objcUtils.addJavaFiles(project, srcFiles, project.j2objcConfig.generatedSourceDir)
```
Comment (brunobowden) [inline j2objc.gradle]: Indent one more space

Comment (brunobowden) [inline j2objc.gradle]: Actually is this just a temporary comment that should be removed entirely?

File: j2objc.gradle [hunk 5]
```diff
@@ -426,7 +427,48 @@ class J2objcTranslateTask extends DefaultTask {
...
+        println "Unchanged files: "+translatedSrcFiles.getFiles().size()
+       
+        
+        if (destDir.exists()) {
+            FileCollection destFiles = project.files(project.fileTree(
+                        dir: destDir, includes: ["**/*.h", "**/*.m"]))
+            
+            // remove translated .h and .m files which has no corresponding .java files anymore
+            destFiles.each {File file ->
+              def nameWithoutExt = file.name.toString().replaceFirst("\\..*","")
```
Comment (brunobowden) [inline j2objc.gradle]: I would prefer that the matching is done on full filenames, rather than comparing the stripped version. Leave this code as is and I'll rework it myself. The code needs to be refactored so that it takes `J2objcUtils.prefixProperties(project)` in to account which mangles the filename.

File: j2objc.gradle [hunk 6]
```diff
@@ -426,7 +427,48 @@ class J2objcTranslateTask extends DefaultTask {
...
+              if (removedFileNames.contains(nameWithoutExt)) {
+                file.delete()
+              }
+            }
+        }
+        
+               
+        // set the srcFiles to the files which need to be translated
+        srcFiles = srcFilesChanged
```
Comment (brunobowden) [inline j2objc.gradle]: Can we eliminate srcFiles and only use srcFilesChanges?

File: j2objc.gradle [hunk 7]
```diff
@@ -468,13 +512,45 @@ class J2objcTranslateTask extends DefaultTask {
...
         if (project.j2objcConfig.filenameCollisionCheck) {
             J2objcUtils.filenameCollisionCheck(srcFiles)
         }
      
         def classPathArg = J2objcUtils.getClassPathArg(
                 project, project.j2objcConfig.translateClassPaths)
-
+          
+        
+        // TODO add translated .class files to classpath for incremental translation
```
Comment (brunobowden) [inline j2objc.gradle]: Is this temporary test code that should be removed?

**Comment by brunobowden** [comment]
@confile - can you help explain this code and what you're trying to do?

```
fileTree(dir: "build/source/apt",
     // TODO @Bruno must be project.j2objcConfig.generatedSourceDir no idea how to set it there
     include: "**/*.java")
```

File: j2objc.gradle (hunk 2 above)
Comment (confile) [inline j2objc.gradle]: done!

File: j2objc.gradle (hunk 5 above)
Comment (confile) [inline j2objc.gradle]: Ok. I leave this up to you.

File: j2objc.gradle (hunk 6 above)
Comment (confile) [inline j2objc.gradle]: No because ``@InputFiles FileCollection srcFiles`` is the input to the tasks containing all source files for this task. The files which have changed are computed after the task has started.

File: j2objc.gradle (hunk 3 above)
Comment (confile) [inline j2objc.gradle]: done

File: j2objc.gradle (hunk 4 above)
Comment (confile) [inline j2objc.gradle]: done

File: j2objc.gradle (hunk 7 above)
Comment (confile) [inline j2objc.gradle]: no this is code your wrote!

**Comment by confile** [comment]
@brunobowden I am done!

**Comment by confile** [comment]
@brunobowden 

```
fileTree(dir: "build/source/apt",
     // TODO @Bruno must be project.j2objcConfig.generatedSourceDir no idea how to set it there
     include: "**/*.java")
```
This code takes the generated source files from an annotation processor which ran in the ``compile-java`` task and appends them to the source files which are the input of the translation task. I have to do it at this point in order to enable the task to track changes.

**Comment by brunobowden** [comment]
I've had the same problem before with trying to setup the correct state from the j2objcConfig object. See the comment from the StackOverflow question I asked. I'm not sure if SO "bigguy" is the same as GitHub @bigguy but he's been very helpful. Still not clear how to wire up the global config but we'll get there...

_... In general, I'd try to stay away from an extension class that contains configuration for a task. There's nothing wrong with that, but you have to do more wiring yourself._ –  bigguy Nov 23 '14 at 19:44

http://stackoverflow.com/questions/27075123/gradle-plugin-best-practices-for-tasks-that-depend-on-extension-objects

File: j2objc.gradle [hunk 8]
```diff
@@ -803,7 +850,10 @@ class j2objc implements Plugin<Project> {
                 srcFiles = files(
                         fileTree(dir: projectDir,
                                 include: "**/*.java",
-                                exclude: project.relativePath(buildDir)))
+                                exclude: project.relativePath(buildDir)) +
+                        fileTree(dir: "build/source/apt",     // TODO @Bruno must be project.j2objcConfig.generatedSourceDir no idea how to set it there
```
Comment (brunobowden) [inline j2objc.gradle]: Put TODO on separate line so it formats better

File: j2objc.gradle [hunk 9]
```diff
@@ -803,7 +850,10 @@ class j2objc implements Plugin<Project> {
...
+                            include: "**/*.java")
```
Comment (brunobowden) [inline j2objc.gradle]: Should be indented 8 spaces as function parameter on separate line

File: j2objc.gradle [hunk 10]
```diff
@@ -829,7 +879,7 @@ class j2objc implements Plugin<Project> {
             }
             
             project.tasks.create(name: 'j2objcXcode', type: J2objcXcodeTask,
-                dependsOn: 'j2objcCopy') {
+                dependsOn: 'j2objcTranslate') {
```
Comment (brunobowden) [inline j2objc.gradle]: Why this change? It should be as it was before.... though I'm thinking of merging the Copy and Xcode tasks together.

File: j2objc.gradle [hunk 11]
```diff
@@ -732,16 +769,26 @@ class J2objcXcodeTask extends DefaultTask {
...
+//                executable "bundle"
+//
+//                args "exec"
+//                args "integratej2objc"
+//                args "integrate_source"
+//                args "-p", "${xcodeProjectDir}"
+//                args "-x", "${xcodeProject}"
+//                args "-s", "${srcFolder}"
+//                args "-g", "${xcodeJ2objcGeneratedDir}"
+//                args "-t", "${xcodeTarget}"
```
Comment (brunobowden) [inline j2objc.gradle]: You should only need to comment out a couple of lines. We'll have to figure out how to make these installations consistent. Let's do that in a separate thread.

**Comment by brunobowden** [comment]
@confile - just a few last nits

File: j2objc.gradle (hunk 10 above)
Comment (confile) [inline j2objc.gradle]: Sorry this was a local change to make the process faster should not be pushed I change it back

**Comment by confile** [comment]
@brunobowden I am done!

File: j2objc.gradle (hunk 1 above)
Comment (brunobowden) [inline j2objc.gradle]: Watch out for the trailing spaces. I'll merge this for now and correct it afterwards.

File: j2objc.gradle [hunk 12]
```diff
@@ -741,7 +778,7 @@ class J2objcXcodeTask extends DefaultTask {
                 args "-x", "${xcodeProject}"
                 args "-s", "${srcFolder}"
                 args "-g", "${xcodeJ2objcGeneratedDir}"
-                args "-t", "${xcodeTarget}"
+                args "-t", "${xcodeTarget}"
```
Comment (brunobowden) [inline j2objc.gradle]: Watch out for the trailing spaces. I'll merge this for now and correct it afterwards.

**Comment by confile** [comment]
@brunobowden Thanks Bruno. Question: Do you have any idea how we could add the already translated classes which have not changed to the ``-classpath`` of the ``j2objc`` command. I tried it here: 
https://github.com/confile/j2objc-gradle/commit/af138abd9a986c2c27b895d6193dfed0cbf79653#diff-cd56e928e1096973b671c75e5e0c6073R525

But when running this code I get the error: ``warning: error in opening zip file``.

**Comment by brunobowden** [comment]
I might be misunderstanding what you're looking but j2objc doesn't yet
translate class files.
